### PR TITLE
Spike/consensus slashing

### PR DESCRIPTION
**Pull request by ZenGround0**
### Motivation
We need to slash consensus faults for a secure protocol.  This PR adds consensus fault detection into the syncing pipeline.  It only detects and emits the faults which are currently dropped.  Handling slashing will come in another PR and is currently blocked on a state machine function that actually executes the penalty on chain.

### Proposed changes
1) New consensus fault detector object + unit tests
2) Integrate construction of fault detector into the syncer submodule and add glue for connecting detected faults to a to-be-written slasher within the start method.
3) Integrate fault detection calls into the syncer.

(3) could happen more decoupled from the syncer by adding in a scanning routine on blocks that have been recently added to the store.  Calling inline in `syncOne` was the simplest way to satisfy needed preconditions of valid parent and verified block signature.  Decoupling detection and actual slashing makes this feel very reasonable.

Doesn't close but addresses part 1 of #3690

### CODE REVIEW

File: cmd/go-filecoin/message_daemon_test.go [hunk 1]
```diff
@@ -126,6 +127,7 @@ func TestMessageSendBlockGasLimit(t *testing.T) {
 	result := struct{ Messages types.TxMeta }{}
 
 	t.Run("when the gas limit is above the block limit, the message fails", func(t *testing.T) {
+		fmt.Printf("WE ARE HERE WE ARE HERE WE ARE HERE\n")
```
Comment (icorderi) [inline cmd/go-filecoin/message_daemon_test.go]: remove

File: internal/pkg/slashing/consensus_slashing.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
+package slashing
+
+import (
+	"github.com/filecoin-project/go-filecoin/internal/pkg/block"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
+)
+
+// CFaultDetector detects consensus faults -- misbehavior conditions where a single
+// party produces multiple blocks at the same time.
+type CFaultDetector struct {
```
Comment (icorderi) [inline internal/pkg/slashing/consensus_slashing.go]: Spell out `Consensus` on the name.

File: internal/pkg/slashing/consensus_slashing.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	// minerIndex tracks witnessed blocks by miner address and epoch
+	minerIndex map[address.Address]map[uint64]*block.Block
+	// sender sends messages on behalf of the slasher
+	faultCh chan ConsensusFault
+}
+
+// ConsensusFault is the information needed to submit a consensus fault
+type ConsensusFault struct {
+	Block1, Block2 *block.Block
```
Comment (icorderi) [inline internal/pkg/slashing/consensus_slashing.go]: what are this blocks? split them and a oneliner with the details.

File: cmd/go-filecoin/message_daemon_test.go (hunk 1 above)
Comment (acruikshank) [inline cmd/go-filecoin/message_daemon_test.go]: Please remove before merging.

File: functional-tests/faucet_test.go [hunk 4]
```diff
@@ -41,7 +42,7 @@ func TestFaucetSendFunds(t *testing.T) {
 	blockTime := time.Second * 5
 
 	// Setup first node, note: Testbed.Name() is the directory
-	genesisTime := time.Unix(123456789, 0)
+	genesisTime := time.Now()
```
Comment (acruikshank) [inline functional-tests/faucet_test.go]: We would like this to be deterministic. I'm assuming this is to debug CI. Unless I'm wrong, please revert before merging.

File: internal/pkg/chainsync/internal/syncer/syncer.go [hunk 5]
```diff
@@ -289,6 +299,15 @@ func (syncer *Syncer) syncOne(ctx context.Context, grandParent, parent, next blo
...
 
+	// Now that the tipset is validated preconditions are satisfied to check
+	// consensus faults
+	for i := 0; i < next.Len(); i++ {
+		err := syncer.faultDetector.CheckBlock(next.At(i), parent)
+		if err != nil {
+			return err
+		}
+	}
+
```
Comment (anorth) [inline internal/pkg/chainsync/internal/syncer/syncer.go]: Since this method proceeds whether or not a fault is detected, I would move this to the end, after storing all state. It's kind of an aside, could even be async.

File: internal/pkg/testhelpers/output.go [hunk 6]
```diff
@@ -112,6 +113,7 @@ func (o *CmdOutput) AssertSuccess() *CmdOutput {
 // AssertFail asserts that the output represents a failed execution, with the error
 // matching the passed in error.
 func (o *CmdOutput) AssertFail(err string) *CmdOutput {
+	fmt.Printf("Assert FAIL\n")
```
Comment (anorth) [inline internal/pkg/testhelpers/output.go]: Remove

File: functional-tests/faucet_test.go (hunk 4 above)
Comment (ZenGround0) [reply functional-tests/faucet_test.go]: Unfortunately the specced out timing rules mean that we either need to do this, mock out the clock, or modify the test to allow for a time-of-test dependent number of null blocks between genesis and the first non-null.  This seems like the best tradeoff for a functional test (if it were an integration test mocking out clock would probably be preferred).
